File: src/info/guardianproject/util/Languages.java
package info.guardianproject.util;

import android.annotation.SuppressLint;
import android.annotation.TargetApi;
import android.app.Activity;
import android.content.ContextWrapper;
import android.content.Intent;
import android.content.res.AssetManager;
import android.content.res.Configuration;
import android.content.res.Resources;
import android.os.Build;
import android.text.TextUtils;
import android.util.DisplayMetrics;

import java.util.Collections;
import java.util.LinkedHashSet;
import java.util.Locale;
import java.util.Map;
import java.util.Set;
import java.util.TreeMap;

public class Languages {
    public static final String TAG = "Languages";

    public static final Locale defaultLocale;
    public static final Locale TIBETAN = new Locale("bo");
    static final Locale localesToTest[] = {
            Locale.ENGLISH, Locale.FRENCH, Locale.GERMAN,
            Locale.ITALIAN, Locale.JAPANESE, Locale.KOREAN,
            Locale.TRADITIONAL_CHINESE, Locale.SIMPLIFIED_CHINESE,
            TIBETAN, new Locale("af"), new Locale("am"),
            new Locale("ar"), new Locale("az"), new Locale("bg"),
            new Locale("bn"), new Locale("ca"), new Locale("cs"),
            new Locale("da"), new Locale("el"), new Locale("es"),
            new Locale("et"), new Locale("eu"), new Locale("fa"),
            new Locale("fi"), new Locale("gl"), new Locale("hi"),
            new Locale("hr"), new Locale("hu"), new Locale("hy"),
            new Locale("in"), new Locale("hy"), new Locale("in"),
            new Locale("is"), new Locale("it"), new Locale("iw"),
            new Locale("ka"), new Locale("kk"), new Locale("km"),
            new Locale("kn"), new Locale("ky"), new Locale("lo"),
            new Locale("lt"), new Locale("lv"), new Locale("mk"),
            new Locale("ml"), new Locale("mn"), new Locale("mr"),
            new Locale("ms"), new Locale("my"), new Locale("nb"),
            new Locale("ne"), new Locale("nl"), new Locale("pl"),
            new Locale("pt"), new Locale("rm"), new Locale("ro"),
            new Locale("ru"), new Locale("si"), new Locale("sk"),
            new Locale("sl"), new Locale("sn"), new Locale("sr"),
            new Locale("sv"), new Locale("sw"), new Locale("ta"),
            new Locale("te"), new Locale("th"), new Locale("tl"),
            new Locale("tr"), new Locale("uk"), new Locale("ur"),
            new Locale("uz"), new Locale("vi"), new Locale("zu"),
    };

    private static final String USE_SYSTEM_DEFAULT = "";
    private static final String defaultString = "Use System Default";

    private static Locale locale;
    private static Languages singleton;
    private static Class<?> clazz;
    private static int resId;
    private static Map<String, String> tmpMap = new TreeMap<String, String>();
    private static Map<String, String> nameMap;

    static {
        defaultLocale = Locale.getDefault();
    }

    private Languages(Activity activity) {
        AssetManager assets = activity.getAssets();
        Configuration config = activity.getResources().getConfiguration();
        // Resources() requires DisplayMetrics, but they are only needed for drawables
        DisplayMetrics ignored = new DisplayMetrics();
        activity.getWindowManager().getDefaultDisplay().getMetrics(ignored);
        Resources resources;
        Set<Locale> localeSet = new LinkedHashSet<Locale>();
        for (Locale locale : localesToTest) {
            config.locale = locale;
            resources = new Resources(assets, ignored, config);
            if (!TextUtils.equals(defaultString, resources.getString(resId))
                    || locale.equals(Locale.ENGLISH))
                localeSet.add(locale);
        }
        for (Locale locale : localeSet) {
            if (locale.equals(TIBETAN)) {
                // include English name for devices without Tibetan font support
                tmpMap.put(TIBETAN.getLanguage(), "Tibetan བོད་སྐད།"); // Tibetan
            } else if (locale.equals(Locale.SIMPLIFIED_CHINESE)) {
                tmpMap.put(Locale.SIMPLIFIED_CHINESE.toString(), "中文 (中国)"); // Chinese (China)
            } else if (locale.equals(Locale.TRADITIONAL_CHINESE)) {
                tmpMap.put(Locale.TRADITIONAL_CHINESE.toString(), "中文 (台灣)"); // Chinese (Taiwan)
            } else {
                tmpMap.put(locale.getLanguage(), locale.getDisplayLanguage(locale));
            }
        }

        /* USE_SYSTEM_DEFAULT is a fake one for displaying in a chooser menu. */
        localeSet.add(null);
        tmpMap.put(USE_SYSTEM_DEFAULT, activity.getString(resId));
        nameMap = Collections.unmodifiableMap(tmpMap);
    }

    /**
     * Get the instance of {@link Languages} to work with, providing the
     * {@link Activity} that is will be working as part of, as well as the
     * {@code resId} that has the exact string "Use System Default",
     * i.e. {@code R.string.use_system_default}.
     * <p/>
     * That string resource {@code resId} is also used to find the supported
     * translations: if an included translation has a translated string that
     * matches that {@code resId}, then that language will be included as a
     * supported language.
     *
     * @param clazz the {@link Class} of the default {@code Activity},
     *              usually the main {@code Activity} from where the
     *              Settings is launched from.
     * @param resId the string resource ID to for the string "Use System Default",
     *              e.g. {@code R.string.use_system_default}
     * @return
     */
    public static void setup(Class<?> clazz, int resId) {
        if (Languages.clazz == null) {
            Languages.clazz = clazz;
            Languages.resId = resId;
        } else {
            throw new RuntimeException("Languages singleton was already initialized, duplicate call to Languages.setup()!");
        }
    }

    /**
     * Get the singleton to work with.
     *
     * @param activity the {@link Activity} this is working as part of
     * @return
     */
    public static Languages get(Activity activity) {
        if (singleton == null) {
            singleton = new Languages(activity);
        }
        return singleton;
    }

    //@TargetApi(Build.VERSION_CODES.JELLY_BEAN_MR1)
    @SuppressLint("NewApi")
	public static void setLanguage(final ContextWrapper contextWrapper, String language, boolean refresh) {
        if (locale != null && TextUtils.equals(locale.getLanguage(), language) && (!refresh)) {
            return; // already configured
        } else if (language == null || language == USE_SYSTEM_DEFAULT) {
            locale = defaultLocale;
        } else {
            /* handle locales with the country in it, i.e. zh_CN, zh_TW, etc */
            String localeSplit[] = language.split("_");
            if (localeSplit.length > 1) {
                locale = new Locale(localeSplit[0], localeSplit[1]);
            } else {
                locale = new Locale(language);
            }
        }

        final Resources resources = contextWrapper.getBaseContext().getResources();
        Configuration config = resources.getConfiguration();
        if (Build.VERSION.SDK_INT >= 17) {
            config.setLocale(locale);
        } else {
            config.locale = locale;
        }
        resources.updateConfiguration(config, resources.getDisplayMetrics());
        Locale.setDefault(locale);

    }

    /**
     * Force reload the {@link Activity to make language changes take effect.}
     *
     * @param activity the {@code Activity} to force reload
     */
    public static void forceChangeLanguage(Activity activity) {
        Intent intent = activity.getIntent();
        if (intent == null) // when launched as LAUNCHER
            return;
        intent.addFlags(Intent.FLAG_ACTIVITY_NO_ANIMATION);
        activity.finish();
        activity.overridePendingTransition(0, 0);
        activity.startActivity(intent);
        activity.overridePendingTransition(0, 0);
    }

    /**
     * Return the name of the language based on the locale.
     *
     * @param locale
     * @return
     */
    public String getName(String locale) {
        String ret = nameMap.get(locale);
        // if no match, try to return a more general name (i.e. English for
        // en_IN)
        if (ret == null && locale.contains("_"))
            ret = nameMap.get(locale.split("_")[0]);
        return ret;
    }

    /**
     * Return an array of the names of all the supported languages, sorted to
     * match what is returned by {@link Languages#getSupportedLocales()}.
     *
     * @return
     */
    public String[] getAllNames() {
        return nameMap.values().toArray(new String[nameMap.size()]);
    }

    public int getPosition(Locale locale) {
        String localeName = locale.getLanguage();
        int i = 0;
        for (String key : nameMap.keySet())
            if (TextUtils.equals(key, localeName))
                return i;
            else
                i++;
        return -1;
    }

    /**
     * Get sorted list of supported locales.
     *
     * @return
     */
    public String[] getSupportedLocales() {
        Set<String> keys = nameMap.keySet();
        return keys.toArray(new String[keys.size()]);
    }
}

File: appcore/src/main/java/org/torproject/android/core/Languages.kt
package org.torproject.android.core

import android.annotation.SuppressLint
import android.app.Activity
import android.content.ContextWrapper
import android.content.res.Resources
import android.os.Build
import android.text.TextUtils
import android.util.DisplayMetrics
import java.util.*

class Languages private constructor(activity: Activity) {
    /**
     * Return an array of the names of all the supported languages, sorted to
     * match what is returned by [Languages.supportedLocales].
     *
     * @return
     */
    val allNames: Array<String>
        get() = nameMap.values.toTypedArray()

    val supportedLocales: Array<String>
        get() {
            val keys = nameMap.keys
            return keys.toTypedArray()
        }

    companion object {
        private var defaultLocale: Locale? = null
        val TIBETAN = Locale("bo")
        val localesToTest = arrayOf(
                Locale.ENGLISH, Locale.FRENCH, Locale.GERMAN,
                Locale.ITALIAN, Locale.JAPANESE, Locale.KOREAN,
                Locale.TRADITIONAL_CHINESE, Locale.SIMPLIFIED_CHINESE,
                TIBETAN, Locale("af"), Locale("am"),
                Locale("ar"), Locale("ay"), Locale("az"), Locale("bg"),
                Locale("bn"), Locale("ca"), Locale("cs"),
                Locale("da"), Locale("el"), Locale("es"),
                Locale("et"), Locale("eu"), Locale("fa"),
                Locale("fi"), Locale("gl"), Locale("hi"),
                Locale("hr"), Locale("hu"), Locale("hy"),
                Locale("in"), Locale("hy"), Locale("in"),
                Locale("is"), Locale("it"), Locale("iw"),
                Locale("ka"), Locale("kk"), Locale("km"),
                Locale("kn"), Locale("ky"), Locale("lo"),
                Locale("lt"), Locale("lv"), Locale("mk"),
                Locale("ml"), Locale("mn"), Locale("mr"),
                Locale("ms"), Locale("my"), Locale("nb"),
                Locale("ne"), Locale("nl"), Locale("pl"),
                Locale("pt"), Locale("rm"), Locale("ro"),
                Locale("ru"), Locale("si"), Locale("sk"),
                Locale("sl"), Locale("sn"), Locale("sr"),
                Locale("sv"), Locale("sw"), Locale("ta"),
                Locale("te"), Locale("th"), Locale("tl"),
                Locale("tr"), Locale("uk"), Locale("ur"),
                Locale("uz"), Locale("vi"), Locale("zu"))
        private const val USE_SYSTEM_DEFAULT = ""
        private const val defaultString = "Use System Default"
        private var locale: Locale? = null
        private var singleton: Languages? = null
        private var clazz: Class<*>? = null
        private var resId = 0
        private val tmpMap: MutableMap<String, String> = TreeMap()
        private lateinit var nameMap: Map<String, String>

        /**
         * Get the instance of [Languages] to work with, providing the
         * [Activity] that is will be working as part of, as well as the
         * `resId` that has the exact string "Use System Default",
         * i.e. `R.string.use_system_default`.
         *
         *
         * That string resource `resId` is also used to find the supported
         * translations: if an included translation has a translated string that
         * matches that `resId`, then that language will be included as a
         * supported language.
         *
         * @param clazz the [Class] of the default `Activity`,
         * usually the main `Activity` from where the
         * Settings is launched from.
         * @param resId the string resource ID to for the string "Use System Default",
         * e.g. `R.string.use_system_default`
         * @return
         */
        @JvmStatic
        fun setup(clazz: Class<*>?, resId: Int) {
            defaultLocale = Locale.getDefault()
            if (Companion.clazz == null) {
                Companion.clazz = clazz
                Companion.resId = resId
            } else {
                throw RuntimeException("Languages singleton was already initialized, duplicate call to Languages.setup()!")
            }
        }

        /**
         * Get the singleton to work with.
         *
         * @param activity the [Activity] this is working as part of
         * @return
         */
        @JvmStatic
        operator fun get(activity: Activity): Languages? {
            if (singleton == null) {
                singleton = Languages(activity)
            }
            return singleton
        }

        @JvmStatic
        @SuppressLint("NewApi")
        fun setLanguage(contextWrapper: ContextWrapper, language: String?, refresh: Boolean) {
            locale = if (locale != null && TextUtils.equals(locale!!.language, language) && !refresh) {
                return  // already configured
            } else if (language == null || language === USE_SYSTEM_DEFAULT) {
                defaultLocale
            } else {
                /* handle locales with the country in it, i.e. zh_CN, zh_TW, etc */
                val localeSplit = language.split("_".toRegex()).toTypedArray()
                if (localeSplit.size > 1) {
                    Locale(localeSplit[0], localeSplit[1])
                } else {
                    Locale(language)
                }
            }
            setLocale(contextWrapper, locale)
        }

        private fun setLocale(contextWrapper: ContextWrapper, locale: Locale?) {
            val resources = contextWrapper.resources
            val configuration = resources.configuration
            val displayMetrics = resources.displayMetrics
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) {
                configuration.setLocale(locale)
                contextWrapper.applicationContext.createConfigurationContext(configuration)
            } else {
                configuration.locale = locale
                resources.updateConfiguration(configuration, displayMetrics)
            }
        }
    }

    init {
        val assets = activity.assets
        val config = activity.resources.configuration
        // Resources() requires DisplayMetrics, but they are only needed for drawables
        val ignored = DisplayMetrics()
        activity.windowManager.defaultDisplay.getMetrics(ignored)
        var resources: Resources
        val localeSet: MutableSet<Locale> = LinkedHashSet()
        for (locale in localesToTest) {
            resources = Resources(assets, ignored, config)
            if (!TextUtils.equals(defaultString, resources.getString(resId))
                    || locale == Locale.ENGLISH) localeSet.add(locale)
        }
        for (locale in localeSet) {
            if (locale == TIBETAN) {
                // include English name for devices without Tibetan font support
                tmpMap[TIBETAN.toString()] = "Tibetan བོད་སྐད།" // Tibetan
            } else if (locale == Locale.SIMPLIFIED_CHINESE) {
                tmpMap[Locale.SIMPLIFIED_CHINESE.toString()] = "中文 (中国)" // Chinese (China)
            } else if (locale == Locale.TRADITIONAL_CHINESE) {
                tmpMap[Locale.TRADITIONAL_CHINESE.toString()] = "中文 (台灣)" // Chinese (Taiwan)
            } else {
                tmpMap[locale.toString()] = locale.getDisplayLanguage(locale)
            }
        }

        /* USE_SYSTEM_DEFAULT is a fake one for displaying in a chooser menu. */
        // localeSet.add(null);
        // tmpMap.put(USE_SYSTEM_DEFAULT, activity.getString(resId));
        nameMap = Collections.unmodifiableMap(tmpMap)
    }
}